mpisppy/utils/sputils.py

```diff
@@ -122,6 +125,97 @@ def reactivate_objs(scenario_instance):
         obj.activate()
 
 
+def id_from_strings(s, vname, idx=None):
+    """ Python id of a Pyomo (indexed) Var
+    WARNING: Pyomo gives a lot of flexibility to modelers, so this might not always work
+    TBD: maybe query the vardata to find out about its indexes and cast based on their types
+    Args:
+        k (str): scenario name
+        s (ConcreteModel): scenario model
+        vname (str): Var name
+        idx (str): index in into Var
+    Returns:
+        varid (python variable id): the id for the particular Pyomo Var
+    """
+    if not hasattr(s, vname):
+        raise RuntimeError(f"Scenario {k} does not have a Var named {vname}")
+    varattr = getattr(s, vname)
+    if idx is None:
+        return id(varattr)
+
+    # if we are still here, we have an index
+    try:
+        return id(varattr[idx])
+    except:
+        pass
+    if ',' not in idx:
+        try:
+            return id(varattr[int(idx)])
+        except:
+            pass
+    else:
+        try:
+            return id(varattr[tuple(idx.split(","))])
+        except:
+            pass
+        try:
+            parts = idx.split(",")
+            return id(varattr[tuple((int(i) for i in parts))])
+        except:
+            pass
+                      
+    
+    print(f"ERROR for var={vname}, idx={idx}; here a all legal indexes:")
+    print([i for i in varattr])
+    raise RuntimeError(f"Scenario (name={s.name}) Var named {vname} does not have index {idx}")
+
+
+
+def rhos_to_csv(s, filename):
+    """ write the rho values to a csv "fullname", rho
+    Args:
+        s (ConcreteModel): the scenario Pyomo model
+        filenaame (str): file to which to write
+    """
+
+    # This is probably already done somewhere...
+    def _ndn_to_node(ndn):
+        for n in s._mpisppy_node_list:
+            if n.name == ndn:
+                return n
+        raise RuntimeError(f"Cannot find {ndn} in node list for scenario {k}")
+    
+    with open(filename, "w") as f:
+        f.write("fullname,rho\n")
+        for ndn_i, rho in s._mpisppy_model.rho.items():
+            n = _ndn_to_node(ndn_i[0])
+            vdata = n.nonant_vardata_list[ndn_i[1]]
+            fullname = vdata.name
+            f.write(f'"{fullname}",{rho._value}\n')
```

Comment:
We should use `_mpisppy_data.nonant_indices`: https://github.com/Pyomo/mpi-sppy/blob/main/mpisppy/spbase.py#L292

Reply:
don.